include/libsemigroups/digraph.hpp

```diff
@@ -3322,5 +3322,162 @@ namespace libsemigroups {
     os << "}";
     return os;
   }
+
+  class HopcroftKarp {
+   private:
+    detail::Duf<> _uf;
+
+   public:
+    HopcroftKarp() = default;
+
+    // Change d in-place to contain the quotient of d1 (or d2) by the partition
+    // returned by the HopcroftKarp Algorithm.
+    template <typename T>
+    detail::Duf<> const& operator()(ActionDigraph<T> const&              d1,
+                                    typename ActionDigraph<T>::node_type p0,
+                                    ActionDigraph<T> const&              d2,
+                                    typename ActionDigraph<T>::node_type q0) {
+      using node_type  = typename ActionDigraph<T>::node_type;
+      using label_type = typename ActionDigraph<T>::label_type;
+
+      auto N1 = d1.number_of_nodes();
+      auto N2 = d2.number_of_nodes();
+
+      if (d1.out_degree() != d2.out_degree()) {
+        LIBSEMIGROUPS_EXCEPTION(
+            "arguments d1 and d2 must have the same out degree");
+      }
+      if (p0 < 0 || p0 >= N1) {
+        LIBSEMIGROUPS_EXCEPTION("argument p0 must be a node of d1");
+      }
+      if (q0 < 0 || q0 >= N2) {
+        LIBSEMIGROUPS_EXCEPTION("argument q0 must be a node of d2");
+      }
+
+      auto M = d1.out_degree();
+      _uf.resize(N1 + N2);
+      _uf.unite(p0, q0 + N1);
+      // 0 .. N1 - 1, N1  .. N1 + N2 -1
+      std::stack<std::pair<node_type, node_type>> stck;
+      stck.emplace(p0, q0 + N1);
+
+      // Traverse d1 and d2, uniting the output vertices at each stage
+      while (!stck.empty()) {
+        node_type q1, q2;
+        std::tie(q1, q2) = stck.top();
+        stck.pop();
+        for (label_type a = 0; a < M; ++a) {
+          node_type r1, r2;
+
+          // Check which digraph q1 and q2 belong to. Vertices with labels from
+          // 0 to N1 correspond to vertices in d1; above N1 corresponds to d2.
+          if (q1 < N1) {
+            r1 = _uf.find(d1.unsafe_neighbor(q1, a));
+          } else {
+            r1 = _uf.find(d2.unsafe_neighbor(q1 - N1, a) + N1);
+          }
+          if (q2 < N1) {
+            r2 = _uf.find(d1.unsafe_neighbor(q2, a));
+          } else {
+            r2 = _uf.find(d2.unsafe_neighbor(q2 - N1, a) + N1);
+          }
+
+          if (r1 != r2) {
+            _uf.unite(r1, r2);
+            stck.emplace(r1, r2);
+          }
+        }
+      }
+
+      // TODO: Decide what the output of this function should be
+      // d1 = quotient_digraph(d1, _uf);
+      return _uf;
+    }
+  };
+
+  template <typename T>
+  ActionDigraph<T> quotient_digraph(ActionDigraph<T> const& d,
```

Comment:
Make this so that it changes the digraph in-place, and then make a version that makes a new copy by calling the in-place function.

Reply:
Resolved with 09fd0ceb79b301220689b2b0d72df53d294c9018